README.md
# useage

##### npm i --save maranda-koa2-session-mysql

### for the complete maranda koa server case, with static, session-mysql, koabody, router

### please visit https://github.com/yu87109/maranda-koa-server-example

### this example is used for ts, if you are js user, if can use as almost the same, just delete the type define.

app.ts
```typescript
import Session, { Sequelize } from 'maranda-koa2-session-mysql'

import Koa from 'koa';

interface SessionData extends Session.DataType{
    name:string,
    id: number,
    stauts:boolean,
    friend: sessionData,
    group: [number,string,sessionData]
}
export interface Ctx extends Session.Ctx<SessionData> { 
    //if have other context
}
const app = new Koa<any, Ctx>();
const sequelize = new Sequelize('xxx', 'xxx', 'xxx', {
    dialect: 'mysql',
    host: 'localhost',
    port: 3306,
})
//you can set the gc probability whith molecula and denominato(this example 5/1000, default 1/100), tableName(custom tablename, default sessions), gcType('auto' or 'manul', if you set it to 'manul, you may do the session gc work by your self), ... 
//you mast ensure that there is not table named 'sessions' or your custom tablename in your database_schema
app.use(Session.middware(
    sequelize,
    {
        gcOpts: { probDenominator: 100, probMolecular: 1, type: 'auto' },
        defaultExpiry: 24 * 60 * 60 * 1000 //if the client close without set the session expiry, the some client will not create session again in 1 day, 
    }
));
app.use((ctx, next) => {
    if (ctx.path == '/'){
        if (ctx.session.isNewRecord) { // if true means that there is no session or the session has been expired of the request 
            ctx.body = `please login`
        }else{
            ...
        }
    }
});
app.use((ctx, next) => {
    if(ctx.path == '/login'){
        const userCode = 'ss',
            passWord = 'sss',
            expiry = 5*24*60*60*1000; //5 days
        ....
        ctx.session.data.friend.id = 'xxx';
        ctx.session.data.id = code;
        ctx.session.expiryTo = new Date(Date.now()+expiry);
        //or you can set expiry as :
        ctx.session.expiry = expiry; //means ctx.session.expiryTo = new Date(ctx.session.createAt.getTime() + expiry)
        // if you do not set expiry or the expiryTo, we set cookie by defult, means when you close the window, the session cookie will be deleted
        // if you do save session data manully, like 'ctx.session.save()', you must set the cookies by your self, like 'ctx.cookie.set(...)'
        await next();
        //do not set your session data after next, because it will never work only if you do save session data munully, like 'ctx.session.save()', and then set the cookies by your self
        ....
    }
});
app.use((ctx, next) => {
    if (ctx.path == '/logout' && !ctx.session.isNewRecord) {
        await ctx.session.destroy();
    }
});

app.listen(80);
```


---

[for more Versions, click see the changelog](./CHANGELOG.md)


CHANGELOG.mD
## V1.0.0
    add useage
## V1.0.1 
    change readme 
## V1.0.2 
    change the git repository.
## V1.0.3 
    change readme 
## V1.0.4 
    fix bug Session.Destory lack await
## V1.0.5 
    fix bug delete ctx.SessKey error 
## V1.0.6 
    change readme 
## V1.0.7 
    redesign apis, export with all sequelize apis, you can use this without install sequelize
## V1.0.8
    modify readme
## V1.0.9
    modify readme
## V2.0.0
    redesign all apis;
## v2.0.1
    add the gc_prob_denominator、gc_prob_molecular in the initOptions replace the gc_probability, your can set the probability of session GC by your self with the denominator and molecular. like gc_prob_denominato = 100 and the gc_prob_molecular =1 , means that the gc_probability = 1/100. Default is 1/100. 
## v2.1.2
    1. add api logger, in the initOptions you can set logger by true/false/function, default is true. Set false means not show GC infomation, Set function you can control the log by your own way like white into a file.
    2. modify the set cookie logic, if in the middleware you do not set the expiry or the ExpiryTo, system will auto set cookie whithout expiries, that means when you close the windows, the cookies destroy   
## v3.0.0
    change the apis to CamelCase;
## v3.0.1
    1. class, namesapce, interface or type defined, function use lowerCamelCase;
    2. variable, attribute, property use UpperCamelCase, if protected start with _, if private start with __;
    3. constant use capital letter and underline 
## v3.0.2
    change the name of InitOptions to SessionInitOptions;
## v3.1.2
    1. re-organization  the api for easier use;
    2. add the generic type for the session data;
    3. set the proxy of the sessionData, for changes; 
## v3.1.4 
    add undefined into the DataTypes
## v3.1.5
    change the error in reademe
## v3.1.8
    fix a bug that session instance change mark is always true.


index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const sequelize_1 = require("sequelize");
class Session extends sequelize_1.Model {
    get expiry() { return this.expiryTo.getTime() - this.createAt.getTime(); }
    set expiry(value) { this.expiryTo = new Date(this.createAt.getTime() + value); }
    gc() {
        return __awaiter(this, void 0, void 0, function* () { return yield Session.destroy({ where: { expiryTo: { [sequelize_1.Op.lt]: new Date() } } }); });
    }
}
(function (Session) {
    function proxyData(session, target) {
        return new Proxy(target, {
            set: (target, key, value) => {
                session.changed('data', true);
                return Reflect.set(target, key, value);
            },
            get: (target, key) => {
                return typeof target[key] !== 'object' ? target[key] : proxyData(session, target[key]);
            }
        });
    }
    function middware(sequelize, initOptions) {
        const { tableName = undefined, gcOpts = { type: 'auto', probMolecular: 1, probDenominator: 100 }, sync = { enable: false, force: false }, sessKey = 'koa2:sess', logger = true, defaultExpiry = 24 * 60 * 60 * 1000 } = initOptions || {};
        Session.init({
            id: { type: sequelize_1.DataTypes.CHAR(36), primaryKey: true, defaultValue: sequelize_1.DataTypes.UUIDV4 },
            data: {
                type: sequelize_1.DataTypes.JSON,
                defaultValue: {},
                get() { return proxyData(this, this.getDataValue('data')); },
                set(value) { this.setDataValue('data', value); }
            },
            expiryTo: { type: sequelize_1.DataTypes.DATE, defaultValue: () => new Date(Date.now() + defaultExpiry) },
            createAt: { type: sequelize_1.DataTypes.DATE, defaultValue: sequelize_1.DataTypes.NOW }
        }, {
            sequelize,
            tableName,
            timestamps: false,
            underscored: true,
            paranoid: false,
        });
        if (sync && sync.enable)
            Session.sync({ force: sync.force }).catch(e => console.log(e));
        return (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            const id = ctx.cookies.get(sessKey);
            if (id) {
                const session = yield Session.findOne({ where: { id } });
                ctx.session = session && session.expiryTo.getTime() > Date.now() ? session : new Session();
            }
            else {
                ctx.session = new Session();
            }
            yield next();
            if (ctx.session.changed()) {
                if (ctx.session.isNewRecord || ctx.session.changed('expiryTo')) {
                    if (ctx.session.changed('expiryTo')) {
                        ctx.cookies.set(sessKey, ctx.session.id, { path: '/', overwrite: true, expires: ctx.session.expiryTo });
                    }
                    else {
                        ctx.cookies.set(sessKey, ctx.session.id, { path: '/', overwrite: true });
                    }
                }
                yield ctx.session.save();
            }
            if (gcOpts.type == 'auto' && Math.random() * Math.abs(gcOpts.probDenominator) <= Math.abs(gcOpts.probMolecular)) {
                if (logger == true)
                    console.log(`[${Date()}]: Auto collect ${yield ctx.session.gc()} session garbage.`);
                if (typeof logger == 'function')
                    logger(yield ctx.session.gc());
            }
        });
    }
    Session.middware = middware;
})(Session || (Session = {}));
exports.default = Session;
__export(require("sequelize"));


index.d.ts
import { Sequelize, Model } from 'sequelize';
import { ParameterizedContext } from 'koa';
declare class Session<T extends Session.DataType> extends Model {
    readonly id: string;
    readonly createAt: Date;
    data: T;
    expiryTo: Date;
    expiry: number;
    gc(): Promise<number>;
}
declare namespace Session {
    interface Ctx<T extends DataType> {
        session: Session<T>;
    }
    interface InitOptions {
        tableName?: string;
        gcOpts: {
            type?: 'auto' | 'manul';
            probDenominator?: number;
            probMolecular?: number;
        };
        sync?: {
            enable: boolean;
            force: boolean;
        };
        sessKey?: string;
        logger?: boolean | ((num: number) => any);
        defaultExpiry?: number;
    }
    interface DataType {
        [key: string]: number | string | boolean | DataType | Array<number | string | boolean | DataType> | undefined;
        [index: number]: number | string | boolean | DataType | Array<number | string | boolean | DataType> | undefined;
    }
    function middware<T extends Ctx<DataType>>(sequelize: Sequelize, initOptions?: InitOptions): (ctx: ParameterizedContext<any, T>, next: () => Promise<any>) => Promise<void>;
}
export default Session;
export * from 'sequelize';


src/index.ts
import { Sequelize, Model, DataTypes, Op } from 'sequelize';
import { ParameterizedContext } from 'koa';

class Session<T extends Session.DataType> extends Model {
  readonly id: string;
  readonly createAt: Date;
  data: T;
  expiryTo: Date;
  get expiry(): number { return this.expiryTo.getTime() - this.createAt.getTime(); }
  set expiry(value: number) { this.expiryTo = new Date(this.createAt.getTime() + value); }
  async gc() { return await Session.destroy({ where: { expiryTo: { [Op.lt]: new Date() } } }); }

}
namespace Session {
  function proxyData<T extends DataType>(session: Session<T>, target: T): T {
    return new Proxy(target, {
      set: (target, key, value) => {
        session.changed('data', true);
        return Reflect.set(target, key, value);
      },
      get: (target, key: string | number) => {
        return typeof target[key] !== 'object' ? target[key] : proxyData(session, <T>target[key]);
      }
    })
  }
  interface Ctx<T extends DataType> { session: Session<T> }
  interface InitOptions {
    tableName?: string,
    gcOpts: {
      type?: 'auto' | 'manul',
      probDenominator?: number,
      probMolecular?: number
    },
    sync?: {
      enable: boolean,
      force: boolean
    },
    sessKey?: string,
    logger?: boolean | ((num: number) => any),
    defaultExpiry?: number
  }
  export interface DataType {
    [key: string]: number | string | boolean | DataType | Array<number | string | boolean | DataType> | undefined
    [index: number]: number | string | boolean | DataType | Array<number | string | boolean | DataType> | undefined
  }
  export function middware<T extends Ctx<DataType>>(sequelize: Sequelize, initOptions?: InitOptions) {
    const {
      tableName = undefined,
      gcOpts = { type: 'auto', probMolecular: 1, probDenominator: 100 },
      sync = { enable: false, force: false },
      sessKey = 'koa2:sess',
      logger = true,
      defaultExpiry = 24 * 60 * 60 * 1000
    } = initOptions || {};
    Session.init(
      {
        id: { type: DataTypes.CHAR(36), primaryKey: true, defaultValue: DataTypes.UUIDV4 },
        data: {
          type: DataTypes.JSON,
          defaultValue: {},
          get<T extends DataType>(this: Session<T>) { return proxyData<T>(this, this.getDataValue('data')) },
          set<T extends DataType>(this: Session<T>, value: T) { this.setDataValue('data', value); }
        },
        expiryTo: { type: DataTypes.DATE, defaultValue: () => new Date(Date.now() + defaultExpiry) },
        createAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW }
      },
      {
        sequelize,
        tableName,
        timestamps: false,
        underscored: true,
        paranoid: false,
      }
    )
    if (sync && sync.enable) Session.sync({ force: sync.force }).catch(e => console.log(e));
    return async (ctx: ParameterizedContext<any, T>, next: () => Promise<any>) => {
      const id = ctx.cookies.get(sessKey);
      if (id) {
        const session = await Session.findOne({ where: { id } });
        ctx.session = session && session.expiryTo.getTime() > Date.now() ? session : new Session();
      } else {
        ctx.session = new Session();
      }
      await next();
      if (ctx.session.changed()) {
        if (ctx.session.isNewRecord || ctx.session.changed('expiryTo')) {
          if (ctx.session.changed('expiryTo')) {
            ctx.cookies.set(sessKey, ctx.session.id, { path: '/', overwrite: true, expires: ctx.session.expiryTo })
          } else {
            ctx.cookies.set(sessKey, ctx.session.id, { path: '/', overwrite: true });
          }
        }
        await ctx.session.save();
      }
      if (gcOpts.type == 'auto' && Math.random() * Math.abs(gcOpts.probDenominator!) <= Math.abs(gcOpts.probMolecular!)) {
        if (logger == true) console.log(`[${Date()}]: Auto collect ${await ctx.session.gc()} session garbage.`)
        if (typeof logger == 'function') logger(await ctx.session.gc());
      }
    }
  }
}
export default Session;
export * from 'sequelize';

